src/outputs/ascent.cpp

```diff
@@ -0,0 +1,203 @@
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+//! \file ascent.cpp
+//  \brief Ascent situ visualization and analysis interop
+
+#include <cstdio>
+#include <cstdlib>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+// Parthenon headers
+#include "coordinates/coordinates.hpp"
+#include "defs.hpp"
+#include "globals.hpp"
+#include "mesh/mesh.hpp"
+#include "outputs/outputs.hpp"
+#include "utils/error_checking.hpp"
+
+// Ascent headers
+#ifdef PARTHENON_ENABLE_ASCENT
+#include "ascent.hpp"
+#include "conduit_blueprint.hpp"
+#include "conduit_relay_io.hpp"
+#include "conduit_relay_io_blueprint.hpp"
+#endif // ifdef PARTHENON_ENABLE_ASCENT
+
+namespace parthenon {
+
+//----------------------------------------------------------------------------------------
+//! \fn void AscentOutput:::WriteOutputFile(Mesh *pm)
+//  \brief  Expose mesh and all Cell variables for processing with Ascent
+void AscentOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
+                                   const SignalHandler::OutputSignal signal) {
+#ifndef PARTHENON_ENABLE_ASCENT
+  if (Globals::my_rank == 0) {
+    PARTHENON_WARN("Ascent output requested by input file, but Ascent support not "
+                   "compiled in. Skipping this output type.");
+  }
+#else
+  // reference:
+  // https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#complete-uniform-example
+
+  using conduit::Node;
+
+  // Ascent needs the MPI communicator we are using
+  ascent::Ascent ascent;
+  Node ascent_opts;
+  ascent_opts["mpi_comm"] = MPI_Comm_c2f(MPI_COMM_WORLD);
+  ascent_opts["actions_file"] = pin->GetString(output_params.block_name, "actions_file");
+  // Only publish fields that are used within actions to reduce memory footprint.
+  // A user might need to override this, e.g., in a runtime ascent_options.yaml, if
+  // the required fields cannot be resolved by Ascent.
+  // See https://ascent.readthedocs.io/en/latest/AscentAPI.html#field-filtering
+  // ascent_opts["field_filtering"] = "true";
```

Comment:
This is supposed to work with the latest ascent develop. I plan to test this out tomorrow.

Reply:
I tested it, but decided to not enable it (for now) until the version with the fix are more broadly available (and I didn't want to introduce specific checks and logic for various Ascent versions)